casanovo/denovo/model.py

```diff
@@ -989,6 +989,190 @@ def configure_optimizers(
         return [optimizer], {"scheduler": lr_scheduler, "interval": "step"}
 
 
+class DBSpec2Pep(Spec2Pep):
+    """
+    Inherits Spec2Pep
+
+    Hijacks teacher-forcing implemented in Spec2Pep and
+    uses it to predict scores between a spectra and associated peptide.
+    Input format is .mgf, with comma-separated targets
+    and decoys in the SEQ field. Decoys should have a prefix of "decoy_".
+    """
+
+    num_pairs = None  # Modified to be predict_batch_size from config
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def predict_step(self, batch, *args):
+        batch_res = []
+        for (
+            indexes,
+            t_or_d,
+            peptides,
+            precursors,
+            encoded_ms,
+        ) in self.smart_batch_gen(batch):
+            pred, truth = self.decoder(peptides, precursors, *encoded_ms)
```

Comment:
The decoder was used directly here to speed up the implementation. Since there are many candidates for each spectrum, I encoded a spectrum once and then copied it multiple times as opposed to replicating the spectrum and encoding a list of copied spectra. This makes it such that we don't need to encode a spectrum `num_candidates` times, but rather just once.